### PR TITLE
Add cancelAndRestart cache strategy for shareFromCache

### PR DESCRIPTION
**Pull request by aim2120**
## Description

Adds an additional caching strategy `cancelAndRestart` that allows for some work to cancel existing work before executing fresh. This may be useful for situations where a new request indicates that an existing requests should be invalidated.

Very open to suggestions!

### CODE REVIEW

**Comment by Tyler-Keith-Thompson** [comment]
I think this makes sense. I think that it may be time to approach this similar to how `RetryStrategy` is approached. That kind of flexible design means that contributions like this are easier and that others who have their own unique behaviors can create them without having to submit a PR. 

That can be done in a follow-up MR however. I think that as it stands having both both these is perfectly reasonable.

File: Sources/Afluent/Workers/ShareFromCache.swift [hunk 1]
```diff
@@ -17,22 +19,45 @@ extension AsynchronousUnitOfWork {
...
+                                cache?.clearAsynchronousUnitOfWork(withKey: key)
+                            },
+                            receiveError: { [weak cache] _ in
+                                cache?.clearAsynchronousUnitOfWork(withKey: key)
+                            },
+                            receiveCancel: { [weak cache] in
+                                cache?.clearAsynchronousUnitOfWork(withKey: key)
+                            }
+                        ).share(),
+                        keyedBy: key) as! CachedWork  // force unwrap since share() returns an opaque type
```
Comment (Tyler-Keith-Thompson) [inline Sources/Afluent/Workers/ShareFromCache.swift:40]: Perhaps we just erase to `AnyAsynchronousUnitOfWork` and avoid the force unwrap?

Comment (aim2120) [reply Sources/Afluent/Workers/ShareFromCache.swift:40]: Sure, I could make that change.

Comment (aim2120) [reply Sources/Afluent/Workers/ShareFromCache.swift]: Ok done!

File: Sources/Afluent/Workers/ShareFromCache.swift [hunk 2]
```diff
@@ -8,6 +8,8 @@
 import Foundation
 
 extension AsynchronousUnitOfWork {
+    private typealias CachedWork = Workers.Share<Workers.HandleEvents<Self, Success>, Success>
```
Comment (Tyler-Keith-Thompson) [inline Sources/Afluent/Workers/ShareFromCache.swift:11]: Still necessary?

**Review by Tyler-Keith-Thompson — APPROVED** [review]

File: Sources/Afluent/Workers/ShareFromCache.swift (hunk 2 above)
Comment (aim2120) [reply Sources/Afluent/Workers/ShareFromCache.swift:11]: We're still using it on L41 for the optional cast. Do you think it's fine to just not cast at all? I mean, presumably the hash key should do the differentiating for us, but curious what you think.

Comment (Tyler-Keith-Thompson) [reply Sources/Afluent/Workers/ShareFromCache.swift:11]: Oh yeah, I don't think we need the cast. At least I don't see any obvious reason why it grants extra safety.

Comment (aim2120) [reply Sources/Afluent/Workers/ShareFromCache.swift]: Ok, removed!

**Pull request merged**